Failure on test.algorithms.test_grover_optimizer.TestGroverOptimizer.test_samples_and_raw_samples

<!-- ⚠️ If you do not respect this template, your issue will be closed -->
<!-- ⚠️ Make sure to browse the opened and closed issues -->

### Information

- **Qiskit Optimization version**: 0.1.0 or 0.2.0
- **Python version**: >=3.6
- **Operating system**: MacOS or Ubuntu

### What is the current behavior?
The following test always fails locally on MacOS with Terra/Aer from master or pypi. On the Ubuntu CI, if I build Aer from master there, it works. If I get Aer from pypi, it fails:
```
==============================
Failed 1 tests - output below:
==============================

test.algorithms.test_grover_optimizer.TestGroverOptimizer.test_samples_and_raw_samples
--------------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/home/runner/work/qiskit-optimization/qiskit-optimization/test/algorithms/test_grover_optimizer.py", line 193, in test_samples_and_raw_samples
    np.testing.assert_array_almost_equal(result.x, result.samples[0].x)

      File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 1044, in assert_array_almost_equal
    precision=decimal)

      File "/opt/hostedtoolcache/Python/3.6.13/x64/lib/python3.6/site-packages/numpy/testing/_private/utils.py", line 840, in assert_array_compare
    raise AssertionError(msg)

    AssertionError: 
Arrays are not almost equal to 6 decimals

Mismatched elements: 1 / 2 (50%)
Max absolute difference: 1.
Max relative difference: 1.
 x: array([0., 0.])
 y: array([1., 0.])
```


### Steps to reproduce the problem

Run test: test.algorithms.test_grover_optimizer.TestGroverOptimizer.test_samples_and_raw_samples

### What is the expected behavior?

It should work

### Suggested solutions

--- a/qiskit_optimization/algorithms/grover_optimizer.py
+++ b/qiskit_optimization/algorithms/grover_optimizer.py
@@ -161,8 +161,7 @@ class GroverOptimizer(OptimizationAlgorithm):
         problem_init = deepcopy(problem_)
 
         # convert to minimization problem
-        sense = problem_.objective.sense
-        if sense == problem_.objective.Sense.MAXIMIZE:
+        if problem_.objective.sense == problem_.objective.Sense.MAXIMIZE:
             problem_.objective.sense = problem_.objective.Sense.MINIMIZE
             problem_.objective.constant = -problem_.objective.constant
             for i, val in problem_.objective.linear.to_dict().items():
@@ -176,7 +175,7 @@ class GroverOptimizer(OptimizationAlgorithm):
         optimum_key = math.inf
         optimum_value = math.inf
         threshold = 0
-        n_key = len(problem_.variables)
+        n_key = self._num_key_qubits
         n_value = self._num_value_qubits
 
         # Variables for tracking the solutions encountered.
@@ -211,6 +210,7 @@ class GroverOptimizer(OptimizationAlgorithm):
                 # Determine the number of rotations.
                 loops_with_no_improvement += 1
                 rotation_count = int(np.ceil(algorithm_globals.random.uniform(0, m - 1)))
+                #rotation_count = 7
                 rotations += rotation_count
                 # Apply Grover's Algorithm to find values below the threshold.
                 # TODO: Utilize Grover's incremental feature - requires changes to Grover.
@@ -220,8 +220,6 @@ class GroverOptimizer(OptimizationAlgorithm):
                 grover = Grover()
                 circuit = grover.construct_circuit(problem=amp_problem,
                                                    power=rotation_count, measurement=measurement)
-
-                # Get the next outcome.
                 outcome = self._measure(circuit)
                 k = int(outcome[0:n_key], 2)
                 v = outcome[n_key:n_key + n_value]
@@ -236,6 +234,19 @@ class GroverOptimizer(OptimizationAlgorithm):
                     logger.info('Current Optimum Value: %s', optimum_value)
                     improvement_found = True
                     threshold = optimum_value
+
+                    # trace out work qubits and store samples
+                    if self._quantum_instance.is_statevector:  # type: ignore
+                        indices = list(range(n_key, len(outcome)))
+                        rho = partial_trace(self._circuit_results, indices)
+                        self._circuit_results = np.diag(rho.data) ** 0.5
+                    else:
+                        self._circuit_results = {i[0:n_key]: v for i,
+                                                v in self._circuit_results.items()}
+
+                    raw_samples = self._eigenvector_to_solutions(self._circuit_results, problem_init)
+                    raw_samples.sort(key=lambda x: problem_.objective.sense.value * x.fval)
+                    samples = self._interpret_samples(problem, raw_samples, self._converters)
                 else:
                     # Using Durr and Hoyer method, increase m.
                     m = int(np.ceil(min(m * 8 / 7, 2 ** (n_key / 2))))
@@ -250,19 +261,6 @@ class GroverOptimizer(OptimizationAlgorithm):
                             len(keys_measured) == num_solutions or rotations >= max_rotations:
                         improvement_found = True
                         optimum_found = True
-
-                # trace out work qubits
-                if self._quantum_instance.is_statevector:  # type: ignore
-                    indices = list(range(n_key, len(outcome)))
-                    rho = partial_trace(self._circuit_results, indices)
-                    self._circuit_results = np.diag(rho.data) ** 0.5
-                else:
-                    self._circuit_results = {i[0:n_key]: v for i,
-                                             v in self._circuit_results.items()}
-
-                raw_samples = self._eigenvector_to_solutions(self._circuit_results, problem_init)
-                raw_samples.sort(key=lambda x: problem_.objective.sense.value * x.fval)
-                samples = self._interpret_samples(problem, raw_samples, self._converters)
 
                 # Track the operation count.
                 operations = circuit.count_ops()
@@ -313,7 +311,7 @@ class GroverOptimizer(OptimizationAlgorithm):
             state = result.get_counts(qc)
             shots = self.quantum_instance.run_config.shots
             hist = {key[::-1]: val / shots for key, val in state.items() if val > 0}
-            self._circuit_results = {b[::-1]: (v / shots) ** 0.5 for (b, v) in state.items()}
+            self._circuit_results = {b[::-1]: np.sqrt(v / shots) for (b, v) in state.items()}
         return hist
 
     @staticmethod

--- a/qiskit_optimization/algorithms/grover_optimizer.py
+++ b/qiskit_optimization/algorithms/grover_optimizer.py
@@ -243,9 +243,10 @@ class GroverOptimizer(OptimizationAlgorithm):
                         self._circuit_results = np.diag(rho.data) ** 0.5
                     else:
                         self._circuit_results = {i[0:n_key]: v for i,
-                                                v in self._circuit_results.items()}
+                                                 v in self._circuit_results.items()}
 
-                    raw_samples = self._eigenvector_to_solutions(self._circuit_results, problem_init)
+                    raw_samples = self._eigenvector_to_solutions(self._circuit_results,
+                                                                 problem_init)
                     raw_samples.sort(key=lambda x: problem_.objective.sense.value * x.fval)
                     samples = self._interpret_samples(problem, raw_samples, self._converters)
                 else:

--- a/qiskit_optimization/algorithms/grover_optimizer.py
+++ b/qiskit_optimization/algorithms/grover_optimizer.py
@@ -210,7 +210,6 @@ class GroverOptimizer(OptimizationAlgorithm):
                 # Determine the number of rotations.
                 loops_with_no_improvement += 1
                 rotation_count = int(np.ceil(algorithm_globals.random.uniform(0, m - 1)))
-                #rotation_count = 7
                 rotations += rotation_count
                 # Apply Grover's Algorithm to find values below the threshold.
                 # TODO: Utilize Grover's incremental feature - requires changes to Grover.
@@ -220,6 +219,8 @@ class GroverOptimizer(OptimizationAlgorithm):
                 grover = Grover()
                 circuit = grover.construct_circuit(problem=amp_problem,
                                                    power=rotation_count, measurement=measurement)
+
+                # Get the next outcome.
                 outcome = self._measure(circuit)
                 k = int(outcome[0:n_key], 2)
                 v = outcome[n_key:n_key + n_value]

--- a/test/algorithms/test_grover_optimizer.py
+++ b/test/algorithms/test_grover_optimizer.py
@@ -22,8 +22,7 @@ from qiskit import Aer
 from qiskit.utils import QuantumInstance, algorithm_globals
 from qiskit.algorithms import NumPyMinimumEigensolver
 from qiskit_optimization.algorithms import (GroverOptimizer,
-                                            MinimumEigenOptimizer,
-                                            OptimizationResultStatus)
+                                            MinimumEigenOptimizer)
 from qiskit_optimization.converters import (InequalityToEquality,
                                             IntegerToBinary,
                                             LinearEqualityToPenalty,

--- a/test/algorithms/test_grover_optimizer.py
+++ b/test/algorithms/test_grover_optimizer.py
@@ -22,7 +22,8 @@ from qiskit import Aer
 from qiskit.utils import QuantumInstance, algorithm_globals
 from qiskit.algorithms import NumPyMinimumEigensolver
 from qiskit_optimization.algorithms import (GroverOptimizer,
-                                            MinimumEigenOptimizer)
+                                            MinimumEigenOptimizer,
+                                            OptimizationResultStatus)
 from qiskit_optimization.converters import (InequalityToEquality,
                                             IntegerToBinary,
                                             LinearEqualityToPenalty,
@@ -169,29 +170,29 @@ class TestGroverOptimizer(QiskitOptimizationTestCase):
 
     def test_samples_and_raw_samples(self):
         """Test samples and raw_samples"""
-        model = Model()
-        x_0 = model.binary_var(name='x0')
-        x_1 = model.binary_var(name='x1')
-        x_2 = model.binary_var(name='x2')
-        model.minimize(-x_0+2*x_1-3*x_2-2*x_0*x_2-1*x_1*x_2)
         op = QuadraticProgram()
-        op.from_docplex(model)
-
-        # Get the optimum key and value.
-        n_iter = 10
-
-        q_instance = self.qasm_simulator
-        gmf = GroverOptimizer(6, num_iterations=n_iter, quantum_instance=q_instance)
-        results = gmf.solve(op)
-        self.assertEqual(len(results.samples), 8)
-        self.assertEqual(len(results.raw_samples), 8)
-        self.assertAlmostEqual(sum(s.probability for s in results.samples), 1)
-        self.assertAlmostEqual(sum(s.probability for s in results.raw_samples), 1)
-        self.assertAlmostEqual(min(s.fval for s in results.samples), -6)
-        self.assertAlmostEqual(min(s.fval for s in results.raw_samples), -6)
-        np.testing.assert_array_almost_equal(results.x, results.samples[0].x)
-        self.assertAlmostEqual(results.fval, results.samples[0].fval)
-        self.assertEqual(results.status, results.samples[0].status)
+        op.integer_var(0, 3, 'x')
+        op.binary_var('y')
+        op.minimize(linear={'x': 1, 'y': 2})
+        op.linear_constraint(linear={'x': 1, 'y': 1}, sense='>=', rhs=1, name='xy')
+        opt_sol = 1
+        success = OptimizationResultStatus.SUCCESS
+        algorithm_globals.random_seed = 1
+        grover_optimizer = GroverOptimizer(
+            8, num_iterations=5, quantum_instance=self.qasm_simulator)
+        result = grover_optimizer.solve(op)
+        self.assertEqual(len(result.samples), 8)
+        self.assertEqual(len(result.raw_samples), 32)
+        self.assertAlmostEqual(sum(s.probability for s in result.samples), 1)
+        self.assertAlmostEqual(sum(s.probability for s in result.raw_samples), 1)
+        self.assertAlmostEqual(min(s.fval for s in result.samples), 0)
+        self.assertAlmostEqual(min(s.fval for s in result.samples if s.status == success), opt_sol)
+        self.assertAlmostEqual(min(s.fval for s in result.raw_samples), opt_sol)
+        for sample in result.raw_samples:
+            self.assertEqual(sample.status, success)
+        np.testing.assert_array_almost_equal(result.x, result.raw_samples[0].x[0:2])
+        self.assertAlmostEqual(result.fval, result.raw_samples[0].fval)
+        self.assertEqual(result.status, result.raw_samples[0].status)
 
 
 if __name__ == '__main__':

--- a/test/algorithms/test_grover_optimizer.py
+++ b/test/algorithms/test_grover_optimizer.py
@@ -170,29 +170,30 @@ class TestGroverOptimizer(QiskitOptimizationTestCase):
 
     def test_samples_and_raw_samples(self):
         """Test samples and raw_samples"""
+        model = Model()
+        x_0 = model.binary_var(name='x0')
+        x_1 = model.binary_var(name='x1')
+        x_2 = model.binary_var(name='x2')
+        model.minimize(-x_0+2*x_1-3*x_2-2*x_0*x_2-1*x_1*x_2)
         op = QuadraticProgram()
-        op.integer_var(0, 3, 'x')
-        op.binary_var('y')
-        op.minimize(linear={'x': 1, 'y': 2})
-        op.linear_constraint(linear={'x': 1, 'y': 1}, sense='>=', rhs=1, name='xy')
-        opt_sol = 1
+        op.from_docplex(model)
         success = OptimizationResultStatus.SUCCESS
-        algorithm_globals.random_seed = 1
-        grover_optimizer = GroverOptimizer(
-            5, num_iterations=2, quantum_instance=self.qasm_simulator)
-        result = grover_optimizer.solve(op)
-        self.assertEqual(len(result.samples), 8)
-        self.assertEqual(len(result.raw_samples), 32)
-        self.assertAlmostEqual(sum(s.probability for s in result.samples), 1)
-        self.assertAlmostEqual(sum(s.probability for s in result.raw_samples), 1)
-        self.assertAlmostEqual(min(s.fval for s in result.samples), 0)
-        self.assertAlmostEqual(min(s.fval for s in result.samples if s.status == success), opt_sol)
-        self.assertAlmostEqual(min(s.fval for s in result.raw_samples), opt_sol)
-        for sample in result.raw_samples:
-            self.assertEqual(sample.status, success)
-        np.testing.assert_array_almost_equal(result.x, result.samples[0].x)
-        self.assertAlmostEqual(result.fval, result.samples[0].fval)
-        self.assertEqual(result.status, result.samples[0].status)
+
+        # Get the optimum key and value.
+        n_iter = 10
+
+        q_instance = self.qasm_simulator
+        gmf = GroverOptimizer(6, num_iterations=n_iter, quantum_instance=q_instance)
+        results = gmf.solve(op)
+        self.assertEqual(len(results.samples), 8)
+        self.assertEqual(len(results.raw_samples), 8)
+        self.assertAlmostEqual(sum(s.probability for s in results.samples), 1)
+        self.assertAlmostEqual(sum(s.probability for s in results.raw_samples), 1)
+        self.assertAlmostEqual(min(s.fval for s in results.samples), -6)
+        self.assertAlmostEqual(min(s.fval for s in results.raw_samples), -6)
+        np.testing.assert_array_almost_equal(results.x, results.samples[0].x)
+        self.assertAlmostEqual(results.fval, results.samples[0].fval)
+        self.assertEqual(results.status, results.samples[0].status)
 
 
 if __name__ == '__main__':

--- a/test/algorithms/test_grover_optimizer.py
+++ b/test/algorithms/test_grover_optimizer.py
@@ -177,7 +177,6 @@ class TestGroverOptimizer(QiskitOptimizationTestCase):
         model.minimize(-x_0+2*x_1-3*x_2-2*x_0*x_2-1*x_1*x_2)
         op = QuadraticProgram()
         op.from_docplex(model)
-        success = OptimizationResultStatus.SUCCESS
 
         # Get the optimum key and value.
         n_iter = 10

